kernel/src/engine/default/file_stream.rs

```diff
@@ -86,24 +90,77 @@ pub struct FileStream<F: FileOpener> {
     /// (before reaching the limit) and returns a batch iterator. If the file reader
     /// is not capable of limiting the number of records in the last batch, the file
     /// stream will take care of truncating it.
-    file_reader: F,
+    file_opener: Box<dyn FileOpener>,
     /// The stream state
     state: FileStreamState,
     /// Describes the behavior of the `FileStream` if file opening or scanning fails
     on_error: OnError,
 }
 
-impl<F: FileOpener> FileStream<F> {
-    /// Create a new `FileStream` using the give `FileOpener` to scan underlying files
+impl FileStream {
+    /// Creates a new `FileStream` from a given schema, `FileOpener`, and files list; the files are
+    /// processed asynchronously by the provided `TaskExecutor`. Returns an `Iterator` that consumes
+    /// the results.
+    pub fn new_async_read_iterator<E: TaskExecutor>(
+        task_executor: Arc<E>,
+        schema: ArrowSchemaRef,
+        file_opener: Box<dyn FileOpener>,
+        files: &[FileMeta],
+        readahead: usize,
+    ) -> DeltaResult<FileDataReadResultIterator> {
+        let mut stream = FileStream::new(files.to_vec(), schema, file_opener)?;
+
+        // This channel will become the output iterator
+        // The stream will execute in the background, and we allow up to `readahead`
+        // batches to be buffered in the channel.
+        let (sender, receiver) = std::sync::mpsc::sync_channel(readahead);
+
+        let executor_for_block = task_executor.clone();
+        task_executor.spawn(async move {
+            while let Some(res) = stream.next().await {
+                let sender = sender.clone();
+                let join_res = executor_for_block
+                    .spawn_blocking(move || sender.send(res))
+                    .await;
```

Comment:
tried to wrap my head around this, but just could not figure it out.

Why do we need to span a new task to do the send operation? and a blocking one at that?

Tried to run the tests without it and it worked fine, so I assume it is related to using this for engines which do not which to implement parquet reading themselves as well so we can send it over FFI? The FFI tests I did not run ...

Given the error message @scovich showed me I would assume this is where the error ultimately originated, since we were blocking on a single threaded runtime?

Reply:
The issue is that the channel could be full. Then the send would block, and can cause deadlock.

See #141 for more details.